drivers/coredump/coredump_impl.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Meta Platforms, Inc. and its affiliates.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <debug/coredump.h>
+#include <drivers/coredump.h>
+
+#define DT_DRV_COMPAT zephyr_coredump
+
+struct coredump_config {
+	/* Length of memory_regions array */
+	int length;
+
+	/* Memory regions specified in device tree */
+	int memory_regions[];
+};
+
+struct coredump_data {
+	/* Memory regions registered at run time */
+	sys_slist_t region_list;
+};
+
+static void coredump_impl_dump(const struct device *dev)
+{
+	const struct coredump_config *config = dev->config;
+
+	/*
+	 * Add each memory region specified in device tree to the core dump,
+	 * the memory_regions array should contain two entries per region
+	 * containing the start address and size.
+	 */
+	if ((config->length > 0) && ((config->length % 2) == 0)) {
+		for (int i = 0; i < config->length; i += 2) {
+			uintptr_t start_address = config->memory_regions[i];
+			size_t size = config->memory_regions[i+1];
+
+			coredump_memory_dump(start_address, start_address + size);
+		}
+	}
+
+	struct coredump_data *data = dev->data;
+	sys_snode_t *node;
+
+	/* Add each memory region registered at runtime to the core dump */
+	SYS_SLIST_FOR_EACH_NODE(&data->region_list, node) {
+		struct coredump_mem_region_node *region;
+
+		region = CONTAINER_OF(node, struct coredump_mem_region_node, node);
+		coredump_memory_dump(region->start, region->start + region->size);
+	}
+}
+
+static void coredump_impl_register_memory(const struct device *dev,
+	struct coredump_mem_region_node *region)
+{
+	struct coredump_data *data = dev->data;
+
+	sys_slist_append(&data->region_list, &region->node);
+}
+
+static bool coredump_impl_unregister_memory(const struct device *dev,
+	struct coredump_mem_region_node *region)
+{
+	struct coredump_data *data = dev->data;
+
+	return sys_slist_find_and_remove(&data->region_list, &region->node);
+}
+
+static int coredump_init(const struct device *dev)
+{
+	struct coredump_data *data = dev->data;
+
+	sys_slist_init(&data->region_list);
+	return 0;
+}
+
+static const struct coredump_driver_api coredump_api = {
+	.dump = coredump_impl_dump,
+	.register_memory = coredump_impl_register_memory,
+	.unregister_memory = coredump_impl_unregister_memory,
+};
+
+#define INIT_REGION(node_id, prop, idx) DT_PROP_BY_IDX(node_id, prop, idx),
+
+#define CREATE_COREDUMP_DEVICE(n)                                                \
+	static struct coredump_data coredump_data_##n;                               \
+	static const struct coredump_config coredump_config##n = {                   \
+		COND_CODE_1(DT_NODE_HAS_PROP(DT_INST(n, DT_DRV_COMPAT), memory_regions), \
+		(                                                                        \
+			.length = DT_PROP_LEN(DT_INST(n, DT_DRV_COMPAT), memory_regions),    \
```

Comment:
can you avoid using the `COND_CODE_1` and use `DT_PROP_LEN_OR` instead?

Reply:
I could use it for the .length property, but I would still need COND_CODE_1 wrapping the DT_INST_FOREACH_PROP_ELEM when building the memory_regions array because that property is optional